In [42]:
import pandas as pd
import numpy as np
import feature_engine
from sklearn.model_selection import train_test_split
import seaborn as sns

In [43]:
data = pd.read_csv(r'../../../data/train.csv')
data.head()

Id  Elevation  Aspect  Slope  Horizontal_Distance_To_Hydrology  \
0   1       2596      51      3                               258   
1   2       2590      56      2                               212   
2   3       2804     139      9                               268   
3   4       2785     155     18                               242   
4   5       2595      45      2                               153   

   Vertical_Distance_To_Hydrology  Horizontal_Distance_To_Roadways  \
0                               0                              510   
1                              -6                              390   
2                              65                             3180   
3                             118                             3090   
4                              -1                              391   

   Hillshade_9am  Hillshade_Noon  Hillshade_3pm  ...  Soil_Type32  \
0            221             232            148  ...            0   
1            220             235            151  ...            0   
2            234             238            135  ...            0   
3            238             238            122  ...            0   
4            220             234            150  ...            0   

   Soil_Type33  Soil_Type34  Soil_Type35  Soil_Type36  Soil_Type37  \
0            0            0            0            0            0   
1            0            0            0            0            0   
2            0            0            0            0            0   
3            0            0            0            0            0   
4            0            0            0            0            0   

   Soil_Type38  Soil_Type39  Soil_Type40  Cover_Type  
0            0            0            0           5  
1            0            0            0           5  
2            0            0            0           2  
3            0            0            0           2  
4            0            0            0           5  

[5 rows x 56 columns]

In [44]:
from sklearn.base import BaseEstimator, TransformerMixin

class DistanceTransformer(BaseEstimator, TransformerMixin):
    def __init__(self):
        pass
    
    def fit(self, X, y=None):
        return self
    
    def transform(self, X):
        X = X.copy()
        X['EuclideanDistanceHidroloy'] = np.around(np.sqrt(X['Horizontal_Distance_To_Hydrology'] **2 + X['Vertical_Distance_To_Hydrology'] **2), 4)
        X.drop(['Horizontal_Distance_To_Hydrology', 'Vertical_Distance_To_Hydrology'], axis=1, inplace=True)
        return X

In [45]:
class DropIdentifierFeatures(BaseEstimator, TransformerMixin):
    def __init__(self):
        pass
    
    def fit(self, X, y=None):
        return self
    
    def transform(self, X):
        X = X.copy()
        X.drop('Id', axis=1, inplace=True)
        return X

In [46]:
X_train, X_test, y_train, y_test = train_test_split(data.drop('Cover_Type', axis=1),
                                                    data['Cover_Type'],
                                                    test_size=.2,
                                                    random_state =123)

In [47]:
y_train.value_counts(normalize=True) * 100, y_test.value_counts(normalize=True) * 100

(2   14.70
 7   14.39
 4   14.32
 3   14.31
 6   14.14
 1   14.11
 5   14.03
 Name: Cover_Type, dtype: float64,
 5   15.31
 1   14.98
 6   14.88
 3   14.19
 4   14.15
 7   13.86
 2   12.63
 Name: Cover_Type, dtype: float64)

In [48]:
from feature_engine.discretisation import EqualWidthDiscretiser
from feature_engine.selection  import DropConstantFeatures, DropDuplicateFeatures, DropCorrelatedFeatures

##  Pipeline

In [49]:
from sklearn.pipeline import Pipeline
pipe = Pipeline([
    ('dropuniquefeatures', DropIdentifierFeatures()),
    ('dp', DropConstantFeatures(tol=0.99)),
    ('dd', DropDuplicateFeatures()),
    ('dt', DistanceTransformer()),
    #('ed', EqualWidthDiscretiser(bins=36, variables=['Aspect'])),
    #('edh', EqualWidthDiscretiser(bins=26, variables=['Hillshade_9am', 'Hillshade_Noon', 'Hillshade_3pm'])),
])
new = pipe.fit_transform(X_train)
new.head()

Elevation  Aspect  Slope  Horizontal_Distance_To_Roadways  \
7711        3151      23     19                             2828   
1466        3101     303     14                             3399   
12128       2079     324     32                              330   
6301        2314     121     15                              644   
9822        3164     117      4                              872   

       Hillshade_9am  Hillshade_Noon  Hillshade_3pm  \
7711             206             197            122   
1466             181             233            192   
12128            126             189            193   
6301             244             227            107   
9822             227             236            143   

       Horizontal_Distance_To_Fire_Points  Wilderness_Area1  Wilderness_Area2  \
7711                                 2200                 0                 0   
1466                                  631                 1                 0   
12128                                 908                 0                 0   
6301                                  702                 0                 0   
9822                                 2536                 1                 0   

       ...  Soil_Type24  Soil_Type29  Soil_Type30  Soil_Type31  Soil_Type32  \
7711   ...            1            0            0            0            0   
1466   ...            0            1            0            0            0   
12128  ...            0            0            0            0            0   
6301   ...            0            0            0            0            0   
9822   ...            0            0            0            0            0   

       Soil_Type33  Soil_Type38  Soil_Type39  Soil_Type40  \
7711             0            0            0            0   
1466             0            0            0            0   
12128            0            0            0            0   
6301             0            0            0            0   
9822             0            0            0            0   

       EuclideanDistanceHidroloy  
7711                     1052.45  
1466                      277.00  
12128                     427.20  
6301                        0.00  
9822                        0.00  

[5 rows x 36 columns]

In [50]:
X_train.shape, new.shape

((12096, 55), (12096, 36))

In [51]:
new.columns

Index(['Elevation', 'Aspect', 'Slope', 'Horizontal_Distance_To_Roadways',
       'Hillshade_9am', 'Hillshade_Noon', 'Hillshade_3pm',
       'Horizontal_Distance_To_Fire_Points', 'Wilderness_Area1',
       'Wilderness_Area2', 'Wilderness_Area3', 'Wilderness_Area4',
       'Soil_Type1', 'Soil_Type2', 'Soil_Type3', 'Soil_Type4', 'Soil_Type5',
       'Soil_Type6', 'Soil_Type10', 'Soil_Type11', 'Soil_Type12',
       'Soil_Type13', 'Soil_Type14', 'Soil_Type17', 'Soil_Type22',
       'Soil_Type23', 'Soil_Type24', 'Soil_Type29', 'Soil_Type30',
       'Soil_Type31', 'Soil_Type32', 'Soil_Type33', 'Soil_Type38',
       'Soil_Type39', 'Soil_Type40', 'EuclideanDistanceHidroloy'],
      dtype='object')

In [52]:
tr = DropCorrelatedFeatures(variables=None, method='pearson', threshold=0.80)
new = tr.fit_transform(new)
new.head()

Elevation  Aspect  Slope  Horizontal_Distance_To_Roadways  \
7711        3151      23     19                             2828   
1466        3101     303     14                             3399   
12128       2079     324     32                              330   
6301        2314     121     15                              644   
9822        3164     117      4                              872   

       Hillshade_9am  Hillshade_Noon  Hillshade_3pm  \
7711             206             197            122   
1466             181             233            192   
12128            126             189            193   
6301             244             227            107   
9822             227             236            143   

       Horizontal_Distance_To_Fire_Points  Wilderness_Area1  Wilderness_Area2  \
7711                                 2200                 0                 0   
1466                                  631                 1                 0   
12128                                 908                 0                 0   
6301                                  702                 0                 0   
9822                                 2536                 1                 0   

       ...  Soil_Type24  Soil_Type29  Soil_Type30  Soil_Type31  Soil_Type32  \
7711   ...            1            0            0            0            0   
1466   ...            0            1            0            0            0   
12128  ...            0            0            0            0            0   
6301   ...            0            0            0            0            0   
9822   ...            0            0            0            0            0   

       Soil_Type33  Soil_Type38  Soil_Type39  Soil_Type40  \
7711             0            0            0            0   
1466             0            0            0            0   
12128            0            0            0            0   
6301             0            0            0            0   
9822             0            0            0            0   

       EuclideanDistanceHidroloy  
7711                     1052.45  
1466                      277.00  
12128                     427.20  
6301                        0.00  
9822                        0.00  

[5 rows x 36 columns]

In [53]:
new.columns

Index(['Elevation', 'Aspect', 'Slope', 'Horizontal_Distance_To_Roadways',
       'Hillshade_9am', 'Hillshade_Noon', 'Hillshade_3pm',
       'Horizontal_Distance_To_Fire_Points', 'Wilderness_Area1',
       'Wilderness_Area2', 'Wilderness_Area3', 'Wilderness_Area4',
       'Soil_Type1', 'Soil_Type2', 'Soil_Type3', 'Soil_Type4', 'Soil_Type5',
       'Soil_Type6', 'Soil_Type10', 'Soil_Type11', 'Soil_Type12',
       'Soil_Type13', 'Soil_Type14', 'Soil_Type17', 'Soil_Type22',
       'Soil_Type23', 'Soil_Type24', 'Soil_Type29', 'Soil_Type30',
       'Soil_Type31', 'Soil_Type32', 'Soil_Type33', 'Soil_Type38',
       'Soil_Type39', 'Soil_Type40', 'EuclideanDistanceHidroloy'],
      dtype='object')

In [54]:
new.shape, y_train.shape

((12096, 36), (12096,))

## Comparing model performaces with engineered data vs Not enginereed

In [55]:
from lazypredict.Supervised import LazyClassifier

In [56]:
clf = LazyClassifier(verbose=0,ignore_warnings=True, custom_metric=None)

In [57]:
X_train_t, X_test_t, y_train_t, y_test_t = train_test_split(new, y_train, test_size=.2,random_state =123)
models,predictions = clf.fit(X_train_t, X_test_t, y_train_t, y_test_t )
predictions

 90%|█████████████████████████████████████████████████████████████████████████▊        | 27/30 [01:10<00:05,  1.67s/it]

[13:35:19] WARNING: C:/Users/Administrator/workspace/xgboost-win64_release_1.3.0/src/learner.cc:1061: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'multi:softprob' was changed from 'merror' to 'mlogloss'. Explicitly set eval_metric if you'd like to restore the old behavior.


100%|██████████████████████████████████████████████████████████████████████████████████| 30/30 [01:17<00:00,  2.60s/it]


Accuracy  Balanced Accuracy ROC AUC  F1 Score  \
Model                                                                          
RandomForestClassifier             0.85               0.85    None      0.85   
ExtraTreesClassifier               0.85               0.85    None      0.85   
LGBMClassifier                     0.85               0.85    None      0.84   
XGBClassifier                      0.84               0.85    None      0.84   
BaggingClassifier                  0.83               0.83    None      0.83   
LabelPropagation                   0.80               0.80    None      0.79   
LabelSpreading                     0.80               0.80    None      0.79   
DecisionTreeClassifier             0.77               0.77    None      0.76   
KNeighborsClassifier               0.76               0.76    None      0.75   
SVC                                0.72               0.73    None      0.72   
ExtraTreeClassifier                0.72               0.72    None      0.72   
LogisticRegression                 0.70               0.70    None      0.70   
CalibratedClassifierCV             0.66               0.66    None      0.66   
LinearSVC                          0.66               0.66    None      0.65   
NuSVC                              0.65               0.65    None      0.65   
LinearDiscriminantAnalysis         0.63               0.63    None      0.63   
RidgeClassifier                    0.61               0.61    None      0.61   
RidgeClassifierCV                  0.61               0.61    None      0.61   
SGDClassifier                      0.60               0.60    None      0.59   
BernoulliNB                        0.59               0.60    None      0.59   
NearestCentroid                    0.58               0.59    None      0.58   
Perceptron                         0.56               0.56    None      0.55   
PassiveAggressiveClassifier        0.53               0.54    None      0.51   
GaussianNB                         0.47               0.47    None      0.37   
QuadraticDiscriminantAnalysis      0.43               0.43    None      0.35   
AdaBoostClassifier                 0.41               0.41    None      0.27   
CheckingClassifier                 0.15               0.14    None      0.04   
DummyClassifier                    0.14               0.14    None      0.14   

                               Time Taken  
Model                                      
RandomForestClassifier               1.82  
ExtraTreesClassifier                 1.50  
LGBMClassifier                       1.66  
XGBClassifier                        5.29  
BaggingClassifier                    0.67  
LabelPropagation                     7.77  
LabelSpreading                       9.62  
DecisionTreeClassifier               0.18  
KNeighborsClassifier                 1.85  
SVC                                  3.86  
ExtraTreeClassifier                  0.05  
LogisticRegression                   0.66  
CalibratedClassifierCV              23.53  
LinearSVC                            6.72  
NuSVC                               10.53  
LinearDiscriminantAnalysis           0.10  
RidgeClassifier                      0.05  
RidgeClassifierCV                    0.08  
SGDClassifier                        0.48  
BernoulliNB                          0.04  
NearestCentroid                      0.05  
Perceptron                           0.11  
PassiveAggressiveClassifier          0.18  
GaussianNB                           0.07  
QuadraticDiscriminantAnalysis        0.07  
AdaBoostClassifier                   0.74  
CheckingClassifier                   0.04  
DummyClassifier                      0.05

In [58]:
from sklearn.model_selection import GridSearchCV
from sklearn.ensemble import RandomForestClassifier, ExtraTreesClassifier
from lightgbm import LGBMClassifier

## Random Forest


### Iteration 1

In [70]:
rfpg = {
    'n_estimators': [12,18,24,48,89,100],
    'max_depth': [2,4,6,8,10,12,14,16,18,20],
    'min_samples_split': [5,8,10,15],
}

In [71]:
rfgs = GridSearchCV(RandomForestClassifier(), rfpg, cv=5, n_jobs=-1, verbose=2)

In [72]:
rfgs.fit(new, y_train)

GridSearchCV(cv=5, estimator=RandomForestClassifier(),
             param_grid={'max_depth': [2, 4, 6, 8, 10, 12, 14, 16, 18, 20],
                         'min_samples_split': [5, 8, 10, 15],
                         'n_estimators': [12, 18, 24, 48, 89, 100]})

In [73]:
rfgs.best_params_

{'max_depth': 20, 'min_samples_split': 5, 'n_estimators': 100}

In [74]:
rfgs.best_score_

0.8451550227366681

In [76]:
pd.DataFrame(rfgs.cv_results_)

mean_fit_time  std_fit_time  mean_score_time  std_score_time  \
0             0.06          0.01             0.01            0.00   
1             0.08          0.00             0.01            0.00   
2             0.11          0.01             0.01            0.00   
3             0.20          0.01             0.01            0.00   
4             0.35          0.00             0.02            0.00   
..             ...           ...              ...             ...   
235           0.26          0.01             0.01            0.00   
236           0.34          0.00             0.01            0.00   
237           0.66          0.01             0.03            0.00   
238           1.21          0.01             0.05            0.00   
239           1.38          0.01             0.05            0.00   

    param_max_depth param_min_samples_split param_n_estimators  \
0                 2                       5                 12   
1                 2                       5                 18   
2                 2                       5                 24   
3                 2                       5                 48   
4                 2                       5                 89   
..              ...                     ...                ...   
235              20                      15                 18   
236              20                      15                 24   
237              20                      15                 48   
238              20                      15                 89   
239              20                      15                100   

                                                params  split0_test_score  \
0    {'max_depth': 2, 'min_samples_split': 5, 'n_es...               0.57   
1    {'max_depth': 2, 'min_samples_split': 5, 'n_es...               0.52   
2    {'max_depth': 2, 'min_samples_split': 5, 'n_es...               0.55   
3    {'max_depth': 2, 'min_samples_split': 5, 'n_es...               0.58   
4    {'max_depth': 2, 'min_samples_split': 5, 'n_es...               0.60   
..                                                 ...                ...   
235  {'max_depth': 20, 'min_samples_split': 15, 'n_...               0.83   
236  {'max_depth': 20, 'min_samples_split': 15, 'n_...               0.83   
237  {'max_depth': 20, 'min_samples_split': 15, 'n_...               0.82   
238  {'max_depth': 20, 'min_samples_split': 15, 'n_...               0.83   
239  {'max_depth': 20, 'min_samples_split': 15, 'n_...               0.84   

     split1_test_score  split2_test_score  split3_test_score  \
0                 0.49               0.56               0.54   
1                 0.58               0.55               0.51   
2                 0.50               0.58               0.52   
3                 0.57               0.57               0.58   
4                 0.59               0.56               0.60   
..                 ...                ...                ...   
235               0.82               0.82               0.81   
236               0.82               0.82               0.82   
237               0.83               0.83               0.82   
238               0.84               0.83               0.82   
239               0.83               0.82               0.82   

     split4_test_score  mean_test_score  std_test_score  rank_test_score  
0                 0.60             0.55            0.04              233  
1                 0.54             0.54            0.02              237  
2                 0.55             0.54            0.03              236  
3                 0.57             0.57            0.01              227  
4                 0.60             0.59            0.02              217  
..                 ...              ...             ...              ...  
235               0.83             0.82            0.01               59  
236               0.83             0.83            0.01               46  
237           

As the gridsearch foun that the min_samples_split was the lower value, and n estimators an max depth where the highest another gridsearch can be created with different hyperparameters

### Iteration 2

In [78]:
rfpg_2 = {
    'n_estimators': [100, 150, 200],
    'max_depth': [20,40,60,80],
    'min_samples_split': [1, 2, 3, 4, 5],
}
rfgs2 = GridSearchCV(RandomForestClassifier(), rfpg_2, cv=5, n_jobs=-1, verbose=2)
rfgs2.fit(new, y_train)


Fitting 5 folds for each of 60 candidates, totalling 300 fits


[Parallel(n_jobs=-1)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=-1)]: Done  25 tasks      | elapsed:   10.6s
[Parallel(n_jobs=-1)]: Done 146 tasks      | elapsed:  1.1min
[Parallel(n_jobs=-1)]: Done 300 out of 300 | elapsed:  2.3min finished


{'max_depth': 20, 'min_samples_split': 5, 'n_estimators': 100}
0.8451550227366681


In [80]:
print(rfgs2.best_params_)
print(rfgs2.best_score_)

{'max_depth': 80, 'min_samples_split': 2, 'n_estimators': 200}
0.8547448744273127


In [82]:
results2 = pd.DataFrame(rfgs2.cv_results_)
results2.head()

mean_fit_time  std_fit_time  mean_score_time  std_score_time  \
0           0.10          0.01             0.00            0.00   
1           0.12          0.02             0.00            0.00   
2           0.19          0.01             0.00            0.00   
3           2.88          0.21             0.12            0.01   
4           3.97          0.22             0.18            0.02   

  param_max_depth param_min_samples_split param_n_estimators  \
0              20                       1                100   
1              20                       1                150   
2              20                       1                200   
3              20                       2                100   
4              20                       2                150   

                                              params  split0_test_score  \
0  {'max_depth': 20, 'min_samples_split': 1, 'n_e...                NaN   
1  {'max_depth': 20, 'min_samples_split': 1, 'n_e...                NaN   
2  {'max_depth': 20, 'min_samples_split': 1, 'n_e...                NaN   
3  {'max_depth': 20, 'min_samples_split': 2, 'n_e...               0.85   
4  {'max_depth': 20, 'min_samples_split': 2, 'n_e...               0.85   

   split1_test_score  split2_test_score  split3_test_score  split4_test_score  \
0                NaN                NaN                NaN                NaN   
1                NaN                NaN                NaN                NaN   
2                NaN                NaN                NaN                NaN   
3               0.85               0.85               0.84               0.86   
4               0.85               0.84               0.84               0.86   

   mean_test_score  std_test_score  rank_test_score  
0              NaN             NaN               60  
1              NaN             NaN               50  
2              NaN             NaN               49  
3             0.85            0.01               32  
4             0.85            0.01               21

#### Conclusion Iteration 2
As the result was that both n_estimator, and max_depth was at the maximum of the Grid i will increase those values

### Iteration 3


In [83]:
rfpg_3 = {
    'n_estimators': [200, 400, 600],
    'max_depth': [80, 160, 240],
}
rfgs3 = GridSearchCV(RandomForestClassifier(), rfpg_3, cv=5, n_jobs=-1, verbose=2)
rfgs3.fit(new, y_train)

Fitting 5 folds for each of 9 candidates, totalling 45 fits


[Parallel(n_jobs=-1)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=-1)]: Done  25 tasks      | elapsed:   42.4s
[Parallel(n_jobs=-1)]: Done  45 out of  45 | elapsed:  1.3min finished


GridSearchCV(cv=5, estimator=RandomForestClassifier(), n_jobs=-1,
             param_grid={'max_depth': [80, 160, 240],
                         'n_estimators': [200, 400, 600]},
             verbose=2)

In [85]:
print(rfgs3.best_params_)
print(rfgs3.best_score_)

{'max_depth': 160, 'n_estimators': 600}
0.8554062706056392


In [86]:
pd.DataFrame(rfgs3.cv_results_)

mean_fit_time  std_fit_time  mean_score_time  std_score_time  \
0           5.96          0.11             0.25            0.04   
1          12.39          0.66             0.48            0.05   
2          18.39          0.58             0.77            0.11   
3           6.12          0.71             0.22            0.02   
4          11.42          0.50             0.51            0.04   
5          16.98          0.29             0.78            0.04   
6           5.46          0.14             0.23            0.02   
7          11.02          0.12             0.48            0.02   
8          14.58          0.93             0.53            0.07   

  param_max_depth param_n_estimators                                   params  \
0              80                200   {'max_depth': 80, 'n_estimators': 200}   
1              80                400   {'max_depth': 80, 'n_estimators': 400}   
2              80                600   {'max_depth': 80, 'n_estimators': 600}   
3             160                200  {'max_depth': 160, 'n_estimators': 200}   
4             160                400  {'max_depth': 160, 'n_estimators': 400}   
5             160                600  {'max_depth': 160, 'n_estimators': 600}   
6             240                200  {'max_depth': 240, 'n_estimators': 200}   
7             240                400  {'max_depth': 240, 'n_estimators': 400}   
8             240                600  {'max_depth': 240, 'n_estimators': 600}   

   split0_test_score  split1_test_score  split2_test_score  split3_test_score  \
0               0.86               0.85               0.86               0.84   
1               0.86               0.85               0.85               0.84   
2               0.86               0.86               0.85               0.84   
3               0.86               0.85               0.85               0.84   
4               0.86               0.86               0.84               0.84   
5               0.86               0.86               0.85               0.84   
6               0.86               0.85               0.84               0.84   
7               0.86               0.86               0.84               0.84   
8               0.86               0.86               0.85               0.84   

   split4_test_score  mean_test_score  std_test_score  rank_test_score  
0               0.86             0.85            0.01                5  
1               0.86             0.85            0.01                6  
2               0.87             0.86            0.01                2  
3               0.87             0.85            0.01                4  
4               0.86             0.85            0.01                8  
5               0.86             0.86            0.01                1  
6               0.87             0.85            0.01                9  
7               0.86             0.85            0.01                7  
8               0.87             0.85            0.01                3

####  Conclusion Iteration 3
After checkih the table its posible to see that there 2 posible solutions with a score of 85.54%, lets see if I can increase a little bit more the precision

### Iteration 4

In [87]:
rfpg_4 = {
    'n_estimators': [600,800,1000,1500],
    'max_depth': [160,180,200,220],
}
rfpg_4 = GridSearchCV(RandomForestClassifier(), rfpg_4, cv=5, n_jobs=-1, verbose=2)
rfpg_4.fit(new, y_train)

Fitting 5 folds for each of 16 candidates, totalling 80 fits


[Parallel(n_jobs=-1)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=-1)]: Done  25 tasks      | elapsed:  1.6min
[Parallel(n_jobs=-1)]: Done  80 out of  80 | elapsed:  5.0min finished


GridSearchCV(cv=5, estimator=RandomForestClassifier(), n_jobs=-1,
             param_grid={'max_depth': [160, 180, 200, 220],
                         'n_estimators': [600, 800, 1000, 1500]},
             verbose=2)

In [91]:
print(rfpg_4.best_params_)
print(rfpg_4.best_score_)
pd.DataFrame(rfpg_4.cv_results_).sort_values(['rank_test_score'])

{'max_depth': 160, 'n_estimators': 1500}
0.8565638078708844


mean_fit_time  std_fit_time  mean_score_time  std_score_time  \
3           42.34          0.36             2.57            0.23   
13          22.02          0.97             0.98            0.04   
1           22.60          0.98             1.01            0.05   
2           27.85          0.69             1.32            0.06   
8           17.05          0.77             0.75            0.06   
15          37.73          2.68             1.57            0.21   
11          42.57          0.72             2.42            0.19   
6           26.98          0.83             1.32            0.08   
5           22.45          0.45             1.03            0.04   
14          28.73          1.15             1.41            0.18   
10          27.41          0.51             1.34            0.12   
9           22.05          0.68             0.99            0.06   
7           43.99          1.11             2.41            0.10   
12          16.23          0.25             0.70            0.04   
0           17.46          0.72             0.78            0.06   
4           16.19          0.33             0.73            0.06   

   param_max_depth param_n_estimators  \
3              160               1500   
13             220                800   
1              160                800   
2              160               1000   
8              200                600   
15             220               1500   
11             200               1500   
6              180               1000   
5              180                800   
14             220               1000   
10             200               1000   
9              200                800   
7              180               1500   
12             220                600   
0              160                600   
4              180                600   

                                      params  split0_test_score  \
3   {'max_depth': 160, 'n_estimators': 1500}               0.86   
13   {'max_depth': 220, 'n_estimators': 800}               0.86   
1    {'max_depth': 160, 'n_estimators': 800}               0.86   
2   {'max_depth': 160, 'n_estimators': 1000}               0.86   
8    {'max_depth': 200, 'n_estimators': 600}               0.86   
15  {'max_depth': 220, 'n_estimators': 1500}               0.86   
11  {'max_depth': 200, 'n_estimators': 1500}               0.86   
6   {'max_depth': 180, 'n_estimators': 1000}               0.86   
5    {'max_depth': 180, 'n_estimators': 800}               0.86   
14  {'max_depth': 220, 'n_estimators': 1000}               0.86   
10  {'max_depth': 200, 'n_estimators': 1000}               0.86   
9    {'max_depth': 200, 'n_estimators': 800}               0.86   
7   {'max_depth': 180, 'n_estimators': 1500}               0.86   
12   {'max_depth': 220, 'n_estimators': 600}               0.86   
0    {'max_depth': 160, 'n_estimators': 600}               0.86   
4    {'max_depth': 180, 'n_estimators': 600}               0.86   

    split1_test_score  split2_test_score  split3_test_score  \
3                0.86               0.85               0.84   
13               0.86               0.85               0.85   
1                0.86               0.85               0.84   
2                0.86               0.85               0.85   
8                0.86               0.85               0.85   
15               0.86               0.85               0.84   
11               0.86               0.85               0.84   
6                0.86               0.85               0.84   
5                0.86               0.85               0.84   
14               0.86               0.85               0.84   
10               0.86               0.85               0.85   
9                0.86               0.85               0.84   
7                0.86               0.85               0.84   
12               0.85               0.85               0.85   
0                0.86               0.85               0.84   
4

####  Conclusion Iteration 4
There is no significant difference for the model of Iteration 3, so the Model 3 Is the chossen one for Random Forests


## LGBMClassifier

### Iteration 1

In [89]:
lgbmg_1 = {
    'boosting_type': ['goss', 'dart', 'gbdt'],
    'num_leaves': [10, 20, 40],
    'class_weight': ['balanced']
}
lgbmg_1 = GridSearchCV(LGBMClassifier(), lgbmg_1, cv=5, n_jobs=-1, verbose=2)
lgbmg_1.fit(new, y_train)

Fitting 5 folds for each of 9 candidates, totalling 45 fits


[Parallel(n_jobs=-1)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=-1)]: Done  25 tasks      | elapsed:   12.1s
[Parallel(n_jobs=-1)]: Done  45 out of  45 | elapsed:   18.1s finished


GridSearchCV(cv=5, estimator=LGBMClassifier(), n_jobs=-1,
             param_grid={'boosting_type': ['goss', 'dart', 'gbdt'],
                         'class_weight': ['balanced'],
                         'num_leaves': [10, 20, 40]},
             verbose=2)

In [92]:
print(lgbmg_1.best_params_)
print(lgbmg_1.best_score_)
pd.DataFrame(lgbmg_1.cv_results_).sort_values('rank_test_score')

{'boosting_type': 'gbdt', 'class_weight': 'balanced', 'num_leaves': 40}
0.8530914010638917


mean_fit_time  std_fit_time  mean_score_time  std_score_time  \
8           1.95          0.18             0.19            0.02   
2           2.90          0.09             0.26            0.01   
5           4.63          0.20             0.22            0.04   
7           1.63          0.05             0.18            0.02   
1           1.92          0.07             0.21            0.01   
4           3.30          0.15             0.15            0.00   
6           1.09          0.04             0.15            0.01   
0           1.34          0.12             0.18            0.02   
3           2.50          0.16             0.12            0.00   

  param_boosting_type param_class_weight param_num_leaves  \
8                gbdt           balanced               40   
2                goss           balanced               40   
5                dart           balanced               40   
7                gbdt           balanced               20   
1                goss           balanced               20   
4                dart           balanced               20   
6                gbdt           balanced               10   
0                goss           balanced               10   
3                dart           balanced               10   

                                              params  split0_test_score  \
8  {'boosting_type': 'gbdt', 'class_weight': 'bal...               0.86   
2  {'boosting_type': 'goss', 'class_weight': 'bal...               0.85   
5  {'boosting_type': 'dart', 'class_weight': 'bal...               0.85   
7  {'boosting_type': 'gbdt', 'class_weight': 'bal...               0.85   
1  {'boosting_type': 'goss', 'class_weight': 'bal...               0.84   
4  {'boosting_type': 'dart', 'class_weight': 'bal...               0.83   
6  {'boosting_type': 'gbdt', 'class_weight': 'bal...               0.81   
0  {'boosting_type': 'goss', 'class_weight': 'bal...               0.80   
3  {'boosting_type': 'dart', 'class_weight': 'bal...               0.79   

   split1_test_score  split2_test_score  split3_test_score  split4_test_score  \
8               0.85               0.85               0.85               0.86   
2               0.85               0.84               0.83               0.85   
5               0.84               0.84               0.84               0.85   
7               0.84               0.83               0.83               0.85   
1               0.83               0.83               0.82               0.85   
4               0.82               0.82               0.81               0.83   
6               0.81               0.81               0.81               0.82   
0               0.80               0.81               0.81               0.82   
3               0.78               0.80               0.78               0.80   

   mean_test_score  std_test_score  rank_test_score  
8             0.85            0.00                1  
2             0.84            0.01                2  
5             0.84            0.01                3  
7             0.84            0.01                4  
1             0.83            0.01                5  
4             0.82            0.01                6  
6             0.81            0.01                7  
0             0.81            0.01                8  
3             0.79            0.01                9

#### Conclusion Iteration 1
There is a minor performance in This model Compare with the one selected on RandomForest, lets play a little bit more with its hyperparameters

### Iteration 2

In [93]:
lgbmg_2 = {
    'boosting_type': ['goss', 'dart', 'gbdt'],
    'num_leaves': [40, 80, 120],
    'class_weight': ['balanced']
}
lgbmg_2 = GridSearchCV(LGBMClassifier(), lgbmg_2, cv=5, n_jobs=-1, verbose=2)
lgbmg_2.fit(new, y_train)

Fitting 5 folds for each of 9 candidates, totalling 45 fits


[Parallel(n_jobs=-1)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=-1)]: Done  25 tasks      | elapsed:   19.7s
[Parallel(n_jobs=-1)]: Done  45 out of  45 | elapsed:   34.6s finished


GridSearchCV(cv=5, estimator=LGBMClassifier(), n_jobs=-1,
             param_grid={'boosting_type': ['goss', 'dart', 'gbdt'],
                         'class_weight': ['balanced'],
                         'num_leaves': [40, 80, 120]},
             verbose=2)

In [94]:
print(lgbmg_2.best_params_)
print(lgbmg_2.best_score_)
pd.DataFrame(lgbmg_2.cv_results_).sort_values('rank_test_score')

{'boosting_type': 'gbdt', 'class_weight': 'balanced', 'num_leaves': 120}
0.8628468836586389


mean_fit_time  std_fit_time  mean_score_time  std_score_time  \
8           5.89          0.15             0.54            0.21   
7           4.13          0.16             0.67            0.16   
5           9.07          0.30             0.59            0.12   
6           2.50          0.07             0.24            0.02   
4           7.02          0.22             0.35            0.06   
1           4.59          0.31             0.70            0.11   
2           5.85          0.46             0.72            0.10   
0           2.92          0.26             0.35            0.04   
3           4.90          0.35             0.23            0.03   

  param_boosting_type param_class_weight param_num_leaves  \
8                gbdt           balanced              120   
7                gbdt           balanced               80   
5                dart           balanced              120   
6                gbdt           balanced               40   
4                dart           balanced               80   
1                goss           balanced               80   
2                goss           balanced              120   
0                goss           balanced               40   
3                dart           balanced               40   

                                              params  split0_test_score  \
8  {'boosting_type': 'gbdt', 'class_weight': 'bal...               0.87   
7  {'boosting_type': 'gbdt', 'class_weight': 'bal...               0.87   
5  {'boosting_type': 'dart', 'class_weight': 'bal...               0.86   
6  {'boosting_type': 'gbdt', 'class_weight': 'bal...               0.86   
4  {'boosting_type': 'dart', 'class_weight': 'bal...               0.86   
1  {'boosting_type': 'goss', 'class_weight': 'bal...               0.86   
2  {'boosting_type': 'goss', 'class_weight': 'bal...               0.85   
0  {'boosting_type': 'goss', 'class_weight': 'bal...               0.85   
3  {'boosting_type': 'dart', 'class_weight': 'bal...               0.85   

   split1_test_score  split2_test_score  split3_test_score  split4_test_score  \
8               0.87               0.85               0.86               0.87   
7               0.86               0.85               0.85               0.87   
5               0.86               0.85               0.85               0.86   
6               0.85               0.85               0.85               0.86   
4               0.85               0.85               0.85               0.86   
1               0.85               0.84               0.84               0.86   
2               0.86               0.84               0.85               0.86   
0               0.85               0.84               0.83               0.85   
3               0.84               0.84               0.84               0.85   

   mean_test_score  std_test_score  rank_test_score  
8             0.86            0.01                1  
7             0.86            0.01                2  
5             0.86            0.01                3  
6             0.85            0.00                4  
4             0.85            0.01                5  
1             0.85            0.01                6  
2             0.85            0.01                7  
0             0.84            0.01                8  
3             0.84            0.01                9

### Conclusion Iteration 2
There is an Increase in gbdt booting type, now lets see if we can find a significance improvement. Score: 86.28%

### Iteration 3

In [95]:
lgbmg_3 = {
    'boosting_type': ['goss', 'dart', 'gbdt'],
    'num_leaves': [120, 240, 360],
    'class_weight': ['balanced']
}
lgbmg_3 = GridSearchCV(LGBMClassifier(), lgbmg_3, cv=5, n_jobs=-1, verbose=2)
lgbmg_3.fit(new, y_train)

Fitting 5 folds for each of 9 candidates, totalling 45 fits


[Parallel(n_jobs=-1)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=-1)]: Done  25 tasks      | elapsed:   40.2s
[Parallel(n_jobs=-1)]: Done  45 out of  45 | elapsed:  1.2min finished


GridSearchCV(cv=5, estimator=LGBMClassifier(), n_jobs=-1,
             param_grid={'boosting_type': ['goss', 'dart', 'gbdt'],
                         'class_weight': ['balanced'],
                         'num_leaves': [120, 240, 360]},
             verbose=2)

In [97]:
print(lgbmg_3.best_params_)
print(lgbmg_3.best_score_)
pd.DataFrame(lgbmg_3.cv_results_).sort_values('rank_test_score').head()

{'boosting_type': 'gbdt', 'class_weight': 'balanced', 'num_leaves': 240}
0.8637561112268918


mean_fit_time  std_fit_time  mean_score_time  std_score_time  \
7           9.61          0.33             1.47            0.28   
8          11.14          0.69             0.97            0.20   
6           5.19          0.17             0.55            0.14   
4          15.80          0.59             1.15            0.15   
5          21.03          0.60             1.78            0.14   

  param_boosting_type param_class_weight param_num_leaves  \
7                gbdt           balanced              240   
8                gbdt           balanced              360   
6                gbdt           balanced              120   
4                dart           balanced              240   
5                dart           balanced              360   

                                              params  split0_test_score  \
7  {'boosting_type': 'gbdt', 'class_weight': 'bal...               0.87   
8  {'boosting_type': 'gbdt', 'class_weight': 'bal...               0.87   
6  {'boosting_type': 'gbdt', 'class_weight': 'bal...               0.87   
4  {'boosting_type': 'dart', 'class_weight': 'bal...               0.86   
5  {'boosting_type': 'dart', 'class_weight': 'bal...               0.86   

   split1_test_score  split2_test_score  split3_test_score  split4_test_score  \
7               0.86               0.86               0.85               0.87   
8               0.86               0.86               0.85               0.87   
6               0.87               0.85               0.86               0.87   
4               0.86               0.86               0.85               0.86   
5               0.86               0.86               0.85               0.87   

   mean_test_score  std_test_score  rank_test_score  
7             0.86            0.01                1  
8             0.86            0.01                2  
6             0.86            0.01                3  
4             0.86            0.01                4  
5             0.86            0.01                5

#### Conclusion Iteration 3
This model seems to have a better Performance

##### 86.37

## Prediction

### Feature Engineering of Test 

In [100]:
transformed_test = pipe.transform(X_test)
transformed_test = tr.transform(transformed_test)

In [103]:
prediction = lgbmg_3.predict(transformed_test)

In [107]:
from sklearn.metrics import accuracy_score, confusion_matrix

### Confusion Matrix

In [118]:
lgbmg_3.score(transformed_test, y_test)

0.8736772486772487

In [108]:
accuracy_score(y_test, prediction)

0.8736772486772487

In [111]:
confusion_matrix(y_test, prediction)

array([[342,  74,   0,   0,   9,   3,  25],
       [ 57, 271,  12,   0,  34,   6,   2],
       [  0,   5, 366,  13,   1,  44,   0],
       [  0,   0,  11, 417,   0,   0,   0],
       [  0,  14,   8,   0, 434,   7,   0],
       [  0,   4,  26,  11,   2, 407,   0],
       [ 13,   1,   0,   0,   0,   0, 405]], dtype=int64)

## Checking with test of kaggle

In [114]:
final_test = pd.read_csv(r'../../../data/test.csv')
final_test.head()

Id  Elevation  Aspect  Slope  Horizontal_Distance_To_Hydrology  \
0  15121       2680     354     14                                 0   
1  15122       2683       0     13                                 0   
2  15123       2713      16     15                                 0   
3  15124       2709      24     17                                 0   
4  15125       2706      29     19                                 0   

   Vertical_Distance_To_Hydrology  Horizontal_Distance_To_Roadways  \
0                               0                             2684   
1                               0                             2654   
2                               0                             2980   
3                               0                             2950   
4                               0                             2920   

   Hillshade_9am  Hillshade_Noon  Hillshade_3pm  ...  Soil_Type31  \
0            196             214            156  ...            0   
1            201             216            152  ...            0   
2            206             208            137  ...            0   
3            208             201            125  ...            0   
4            210             195            115  ...            0   

   Soil_Type32  Soil_Type33  Soil_Type34  Soil_Type35  Soil_Type36  \
0            0            0            0            0            0   
1            0            0            0            0            0   
2            0            0            0            0            0   
3            0            0            0            0            0   
4            0            0            0            0            0   

   Soil_Type37  Soil_Type38  Soil_Type39  Soil_Type40  
0            0            0            0            0  
1            0            0            0            0  
2            0            0            0            0  
3            0            0            0            0  
4            0            0            0            0  

[5 rows x 55 columns]

In [116]:
final_test_features = pipe.transform(final_test.drop('Cover_Type', axis=1))

KeyError: "['Cover_Type'] not found in axis"

In [16]:
X_train_um, X_test_um, y_train_um, y_test_um = train_test_split(data.drop('Cover_Type', axis=1),
                                                    data['Cover_Type'],
                                                    test_size=.1,
                                                    random_state =123)
clf = LazyClassifier(verbose=0,ignore_warnings=True, custom_metric=None)
models,predictions = clf.fit(X_train, X_test, y_train, y_test)

 90%|█████████████████████████████████████████████████████████████████████████▊        | 27/30 [02:19<00:11,  3.68s/it]

[12:54:36] WARNING: C:/Users/Administrator/workspace/xgboost-win64_release_1.3.0/src/learner.cc:1061: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'multi:softprob' was changed from 'merror' to 'mlogloss'. Explicitly set eval_metric if you'd like to restore the old behavior.


100%|██████████████████████████████████████████████████████████████████████████████████| 30/30 [02:27<00:00,  4.92s/it]


In [17]:
models

Accuracy  Balanced Accuracy ROC AUC  F1 Score  \
Model                                                                          
RandomForestClassifier             0.89               0.88    None      0.89   
LGBMClassifier                     0.89               0.88    None      0.88   
XGBClassifier                      0.88               0.88    None      0.88   
ExtraTreesClassifier               0.88               0.88    None      0.88   
BaggingClassifier                  0.85               0.85    None      0.85   
LabelPropagation                   0.81               0.80    None      0.81   
LabelSpreading                     0.81               0.80    None      0.81   
KNeighborsClassifier               0.79               0.79    None      0.79   
DecisionTreeClassifier             0.79               0.78    None      0.79   
SVC                                0.76               0.76    None      0.76   
ExtraTreeClassifier                0.76               0.75    None      0.75   
LogisticRegression                 0.73               0.73    None      0.73   
CalibratedClassifierCV             0.71               0.71    None      0.71   
LinearSVC                          0.71               0.71    None      0.71   
NuSVC                              0.70               0.70    None      0.70   
LinearDiscriminantAnalysis         0.68               0.67    None      0.68   
RidgeClassifier                    0.66               0.66    None      0.66   
RidgeClassifierCV                  0.66               0.66    None      0.66   
BernoulliNB                        0.65               0.64    None      0.64   
SGDClassifier                      0.64               0.64    None      0.64   
NearestCentroid                    0.63               0.63    None      0.63   
Perceptron                         0.60               0.60    None      0.58   
PassiveAggressiveClassifier        0.48               0.48    None      0.48   
QuadraticDiscriminantAnalysis      0.47               0.47    None      0.39   
GaussianNB                         0.47               0.46    None      0.38   
AdaBoostClassifier                 0.46               0.46    None      0.39   
CheckingClassifier                 0.14               0.14    None      0.04   
DummyClassifier                    0.14               0.14    None      0.14   

                               Time Taken  
Model                                      
RandomForestClassifier               2.38  
LGBMClassifier                       1.78  
XGBClassifier                        6.15  
ExtraTreesClassifier                 1.97  
BaggingClassifier                    1.37  
LabelPropagation                    12.76  
LabelSpreading                      15.38  
KNeighborsClassifier                 2.09  
DecisionTreeClassifier               0.24  
SVC                                  9.12  
ExtraTreeClassifier                  0.08  
LogisticRegression                   0.78  
CalibratedClassifierCV              48.81  
LinearSVC                           12.12  
NuSVC                               28.65  
LinearDiscriminantAnalysis           0.31  
RidgeClassifier                      0.08  
RidgeClassifierCV                    0.14  
BernoulliNB                          0.07  
SGDClassifier                        1.08  
NearestCentroid                      0.07  
Perceptron                           0.29  
PassiveAggressiveClassifier          0.27  
QuadraticDiscriminantAnalysis        0.11  
GaussianNB                           0.09  
AdaBoostClassifier                   1.27  
CheckingClassifier                   0.05  
DummyClassifier                      0.05

[[0, 2, 0, 3], [0, 1, 4, 3], [0, 1, 1, 3]]### Results Iteration 1

The engineered features perform better, at least with a quick check on XGBoosts, and the not engineered features perform well with RandomForest. Nevertheless the time performance in later features is worst than the one with the second models check.